### DISCUSSION TITLE
Don't use limit param in a search function

### DISCUSSION THREAD

**Discussion by Jaroslove** [milvus-io/milvus#27160] [Q&A and General discussion] [answered]
Hi Team, 

I need to find most similar vectors in the db but I need to filter them by distance not by limit param.

Is it possible to not use limit param in a search function because I don't know how many vectors are present in the db
which will be match by my distance criteria? 

Also, in my case I may have limit param more than 16384.

Thank you!

**Comment by yhmo** [comment]
Range Search is the feature you are looking for: https://milvus.io/docs/within_range.md#Conduct-a-Range-Search

Example:

```
import random

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
    Index
)

if __name__ == '__main__':
    connections.connect(host=HOST, port=PORT)
    print(utility.get_server_version())

    collection_name = "bbb"

    if utility.has_collection(collection_name):
        utility.drop_collection(collection_name)

    field_1 = FieldSchema(
        name="id",
        dtype=DataType.INT64,
        is_primary=True,
        auto_id=False,
    )
    field_2 = FieldSchema(
        name="vector",
        dtype=DataType.FLOAT_VECTOR,
        dim=128
    )
    schema = CollectionSchema(fields=[field_1, field_2])
    collection = Collection(name=collection_name, schema=schema)

    num = 50000
    data = [
        [i for i in range(num)],
        [[random.random() for _ in range(128)] for _ in range(num)],
    ]
    collection.insert(data)
    collection.flush()

    index_params = {
        'metric_type': 'L2',
        'index_type': "FLAT",
        'params': {}
    }
    collection.create_index(field_name="vector", index_params=index_params)
    collection.load()

    collection = Collection(name=collection_name)
    target_vecotrs = [[random.random() for _ in range(128)]]

    total_results = []
    inner_radius = 0.0
    outer_radius = 1000.0
    while True:
        print(f"Search from inner radius {inner_radius} to outer radius {outer_radius}")
        results = collection.search(
            data=target_vecotrs,
            anns_field="vector",
            param={"metric_type": "L2", "params": {"range_filter": inner_radius, "radius": outer_radius}},
            limit=16384,
        )
        items = results[0]
        inner_radius = items[len(items)-1].distance

        if len(total_results) == 0:
            total_results.extend(list(items))
            print(f"Results count: {len(items)} in distance range: [{items[0].distance}, {items[len(items)-1].distance}]")
            continue
        elif len(items) > 1:
            total_results.extend(list(items)[1:])
            print(f"Results count: {len(items)-1} in distance range: [{items[1].distance}, {items[len(items)-1].distance}]")
            continue

        if len(items) == 1:
            break

    print(f"Total results count {len(total_results)}")

```

**Reply by Jaroslove** [reply]
Thank you for the answer. 

Probably, I may have a case when my db has a lot of the same vectors. 
In this case the code that you provided to me doesn't work. 
I have slightly changed the code and I got an infite loop.

Please, see the changed code below:



        connections.connect("default", host="localhost", port="19530")
        print(utility.get_server_version())

        collection_name = "bbb"

        if utility.has_collection(collection_name):
            utility.drop_collection(collection_name)

        field_1 = FieldSchema(
            name="id",
            dtype=DataType.INT64,
            is_primary=True,
            auto_id=False,
        )
        field_2 = FieldSchema(
            name="vector",
            dtype=DataType.FLOAT_VECTOR,
            dim=128
        )
        schema = CollectionSchema(fields=[field_1, field_2])
        collection = Collection(name=collection_name, schema=schema)

        vector = [random.random() for _ in range(128)]

        num = 50000
        data = [
            [i for i in range(num)],
            [vector for _ in range(num)],
        ]
        collection.insert(data)
        collection.flush()

        index_params = {
            'metric_type': 'L2',
            'index_type': "FLAT",
            'params': {}
        }
        collection.create_index(field_name="vector", index_params=index_params)
        collection.load()

        collection = Collection(name=collection_name)
        target_vecotrs = [vector]

        total_results = []
        inner_radius = 0.0
        outer_radius = 1000.0
        while True:
            print(f"Search from inner radius {inner_radius} to outer radius {outer_radius}")
            results = collection.search(
                data=target_vecotrs,
                anns_field="vector",
                param={"metric_type": "L2", "params": {"range_filter": inner_radius, "radius": outer_radius}},
                limit=16384,
            )
            items = results[0]
            inner_radius = items[len(items) - 1].distance

            if len(total_results) == 0:
                total_results.extend(list(items))
                print(
                    f"Results count: {len(items)} in distance range: [{items[0].distance}, {items[len(items) - 1].distance}]")
                continue
            elif len(items) > 1:
                total_results.extend(list(items)[1:])
                print(
                    f"Results count: {len(items) - 1} in distance range: [{items[1].distance}, {items[len(items) - 1].distance}]")
                continue

            if len(items) == 1:
                break

        print(f"Total results count {len(total_results)}")

**Comment by Jaroslove** [comment]
Also, I tried to use offset param to skip searched records but I got an error 

status {
  error_code: UnexpectedError
  reason: "offset+topk [32768] is invalid, top k should be in range [1, 16384], but got 32768"
}

****
_Please, see the changed code below:_



        print(utility.get_server_version())

        collection_name = "bbb"

        if utility.has_collection(collection_name):
            utility.drop_collection(collection_name)

        field_1 = FieldSchema(
            name="id",
            dtype=DataType.INT64,
            is_primary=True,
            auto_id=False,
        )
        field_2 = FieldSchema(
            name="vector",
            dtype=DataType.FLOAT_VECTOR,
            dim=128
        )
        schema = CollectionSchema(fields=[field_1, field_2])
        collection = Collection(name=collection_name, schema=schema)

        vector = [random.random() for _ in range(128)]

        num = 50000
        data = [
            [i for i in range(num)],
            [vector for _ in range(num)],
        ]
        collection.insert(data)
        collection.flush()

        index_params = {
            'metric_type': 'L2',
            'index_type': "FLAT",
            'params': {}
        }
        collection.create_index(field_name="vector", index_params=index_params)
        collection.load()

        collection = Collection(name=collection_name)
        target_vecotrs = [vector]

        total_results = []
        inner_radius = 0.0
        outer_radius = 1000.0
        offset = 0
        while True:
            print(f"Search from inner radius {inner_radius} to outer radius {outer_radius}")
            results = collection.search(
                data=target_vecotrs,
                anns_field="vector",
                param={"metric_type": "L2",  "offset": offset, "params": {"range_filter": inner_radius, "radius": outer_radius}},
                limit=16384,
            )
            items = results[0]
            inner_radius = items[len(items) - 1].distance

            if len(total_results) == 0:
                total_results.extend(list(items))
                print(
                    f"Results count: {len(items)} in distance range: [{items[0].distance}, {items[len(items) - 1].distance}]")
                offset += len(total_results)
                continue
            elif len(items) > 1:
                total_results.extend(list(items)[1:])
                print(
                    f"Results count: {len(items) - 1} in distance range: [{items[1].distance}, {items[len(items) - 1].distance}]")
                offset += len(total_results)
                continue

            if len(items) == 1:
                break

        print(f"Total results count {len(total_results)}")

**Reply by yhmo** [reply, marked as answer]
If there are lots of duplicate vectors, the only way to search all of them is adding a hidden configuration to the milvus.yaml:
```
quotaAndLimits:
  limits:
    topK: 1000000
```

Set quotaAndLimits.limits.topK=1000000 allows the search() to return at most 1000000 items.

(In my opinion, 50000 duplicated vectors in a db is meaningless, the application design is not reasonable)
